pkg/util/purge/activedirectory.go

```diff
@@ -0,0 +1,158 @@
+package purge
+
+// Copyright (c) Microsoft Corporation.
+// Licensed under the Apache License 2.0.
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"regexp"
+	"strings"
+
+	"github.com/Azure/azure-sdk-for-go/services/preview/authorization/mgmt/2018-09-01-preview/authorization"
+	"github.com/Azure/go-autorest/autorest"
+)
+
+func (rc *ResourceCleaner) CleanAAD(ctx context.Context) error {
+	roleAssignments, err := rc.listRoleAssignmentsForVnets(ctx)
+	if err != nil {
+		return err
+	}
+
+	return rc.cleanAAD(ctx, roleAssignments)
+}
+
+// getAROClusters lists all resource groups and filter those, that match the pattern: 'aro-[a-z0-9]{8}'
+func (rc *ResourceCleaner) getAROClusers(ctx context.Context) map[string]bool {
+	return nil
+}
+
+func (rc *ResourceCleaner) listRoleAssignmentsForVnets(ctx context.Context) ([]authorization.RoleAssignment, error) {
+	roleAssignmentsResourceGroups := []string{"v4-westeurope", "v4-eastus", "v4-australiasoutheast"}
+	var filteredRoleAssignments []authorization.RoleAssignment
+
+	for _, rg := range roleAssignmentsResourceGroups {
+		rc.log.Debugf("rg: %s", rg)
+		vnets, err := rc.vnetscli.List(ctx, rg)
+		if err != nil {
+			return nil, err
+		}
+
+		for _, vnet := range vnets {
+			scope := fmt.Sprintf("/subscriptions/%s/resourceGroups/%s/providers/Microsoft.Network/virtualNetworks/%s", rc.subscriptionID, rg, *vnet.Name)
+			roleAssignments, err := rc.roleassignmentcli.ListForScope(ctx, scope, "")
+			if err != nil {
+				return nil, err
+			}
+
+			for _, rs := range roleAssignments {
+				if rs.PrincipalType != authorization.ServicePrincipal {
+					// work only on application types
+					continue
+				}
+
+				filteredRoleAssignments = append(filteredRoleAssignments, rs)
+				rc.log.Debugf("RG: %s, roleAssignemnt: %s\n", rg, rs)
+			}
+		}
+	}
+
+	return filteredRoleAssignments, nil
+}
+
+func (rc *ResourceCleaner) cleanAAD(ctx context.Context, roleAssignments []authorization.RoleAssignment) error {
```

Comment:
Can we have a behaviour documented as part of function documentations?
Like checking back relationship, AD ID URL, cluster existence. Basic logical flow explained.

Reply:
will this do?